File: src/SharpPilot/ToolsStatusConfig.cs
namespace SharpPilot;

using System.Text.Json;

/// <summary>
/// Reads tool enable/disable state from a <c>tools-status.json</c> file
/// located next to the server binary. When the file is missing or unreadable,
/// all tools are treated as enabled.
/// </summary>
internal static class ToolsStatusConfig
{
    private static readonly string StatusFilePath =
        Path.Combine(AppContext.BaseDirectory, "tools-status.json");

    /// <summary>
    /// Returns <see langword="true"/> when the tool identified by
    /// <paramref name="toolName"/> is enabled (or when the status file
    /// is absent / does not contain the key).
    /// </summary>
    internal static bool IsEnabled(string toolName)
    {
        if (!File.Exists(StatusFilePath))
        {
            return true;
        }

        try
        {
            var json = File.ReadAllText(StatusFilePath);
            using var doc = JsonDocument.Parse(json);

            if (doc.RootElement.TryGetProperty(toolName, out var value)
                && value.ValueKind == JsonValueKind.False)
            {
                return false;
            }

            return true;
        }
        catch (Exception ex) when (ex is IOException or JsonException)
        {
            return true;
        }
    }
}

File: src/SharpPilot.Mcp.Web/src/configuration/tools-status-config.ts
import { readFileSync } from 'node:fs';
import { join } from 'node:path';

export class ToolsStatusConfig {
    private readonly configFilePath: string | undefined;

    constructor(workspacePath?: string) {
        this.configFilePath = workspacePath
            ? join(workspacePath, '.sharppilot.json')
            : undefined;
    }

    isEnabled(toolName: string): boolean {
        if (!this.configFilePath) {
            return true;
        }

        try {
            const json = readFileSync(this.configFilePath, 'utf-8');
            const doc: unknown = JSON.parse(json);

            if (
                typeof doc === 'object' && doc !== null
                && 'tools' in doc && typeof doc.tools === 'object' && doc.tools !== null
                && 'disabledTools' in doc.tools && Array.isArray(doc.tools.disabledTools)
            ) {
                return !doc.tools.disabledTools.includes(toolName);
            }

            return true;
        } catch {
            return true;
        }
    }
}
